### ISSUE TITLE
Weekly snapshots not being kept as expected

### ISSUE THREAD

**Issue by shaaati** [Patryk27/lxd-snapper#13]
I have been running lxd-snapper for about three months now.
My configuration looks like this:

```hooks:
  # start of process
  on-backup-started: 'echo "[*] `date` on-backup-started" >> resolute-backup_snapper.log'
  # check if container is running after snapshotting and stop it (I've had this for quite some time when copying snapshots to remotes, don't know why but since then I manually stop whatever might be running)
  on-snapshot-created: 'if [ `lxc ls {{ instanceName }} --column=s --format=csv` = "RUNNING" ]; then lxc stop {{ instanceName }}; echo "[!] `date` Container {{ instanceName }} was stopped, should not be running" >> resolute-backup_snapper.log; fi'
  # transfer after succesful snapshot as suggested by author of lxd-snapper
  on-instance-backed-up: 'lxc copy --refresh {{ remoteName }}:{{ instanceName }} {{ instanceName }}; echo "[*] `date` Refreshed container {{ instanceName }} from remote host, copied new snapshots" >> resolute-backup_snapper.log'
  # transfer after successful prune as suggested by author of lxd-snapper
  on-instance-pruned: 'lxc copy --refresh {{ remoteName }}:{{ instanceName }} {{ instanceName }}; echo "[*] `date` Refreshed container {{ instanceName }} from remote host, deleted old snapshots" >> resolute-backup_snapper.log'
  # as before, check if container is running after backup and stop it
  on-backup-completed: 'for container in `lxc ls -c n,s --format csv | grep -v STOPPED`; do lxc stop `echo $container | cut -d"," -f1`; echo "[!] `date` Container `echo $container | cut -d"," -f1` was stopped, should not be running" >> resolute-backup_snapper.log; done'

remotes:
  - resolute

policies:
  all-containers:
    keep-daily: 6
    keep-weekly: 3
    keep-monthly: 5
    keep-yearly: 3
```

In the beginning, it seemed mostly fine, even though I could never manage to get more than two weekly snapshots, see, e.g., the output of the run June 20:
```
  - keeping snapshot: lxd-snapper20240619-220005
  - keeping snapshot: lxd-snapper20240618-220002
  - keeping snapshot: lxd-snapper20240617-220002
  - keeping snapshot: lxd-snapper20240616-220003
  - keeping snapshot: lxd-snapper20240615-220007
  - keeping snapshot: lxd-snapper20240614-220002
  - keeping snapshot: lxd-snapper20240613-220002
  - keeping snapshot: lxd-snapper20240612-220003
  - deleting snapshot: lxd-snapper20240611-220002 [ OK ]
  - keeping snapshot: lxd-snapper20240608-220008
  - keeping snapshot: lxd-snapper20240601-220002
  - keeping snapshot: lxd-snapper20240531-220002
```

Somehow, when the last week of the month starts, weekly backups seem to get deleted: (run from June 23):
```
  - keeping snapshot: lxd-snapper20240622-220004
  - keeping snapshot: lxd-snapper20240621-220003
  - keeping snapshot: lxd-snapper20240620-220003
  - keeping snapshot: lxd-snapper20240619-220005
  - keeping snapshot: lxd-snapper20240618-220002
  - keeping snapshot: lxd-snapper20240617-220002
  - keeping snapshot: lxd-snapper20240616-220003
  - keeping snapshot: lxd-snapper20240615-220007
  - keeping snapshot: lxd-snapper20240614-220002
  - keeping snapshot: lxd-snapper20240608-220008
  - deleting snapshot: lxd-snapper20240601-220002 [ OK ]
  - keeping snapshot: lxd-snapper20240531-220002
```

Run from June 25:
```
  - keeping snapshot: lxd-snapper20240624-220006
  - keeping snapshot: lxd-snapper20240623-220005
  - keeping snapshot: lxd-snapper20240622-220006
  - keeping snapshot: lxd-snapper20240621-220005
  - keeping snapshot: lxd-snapper20240620-220005
  - keeping snapshot: lxd-snapper20240619-220007
  - keeping snapshot: lxd-snapper20240618-220004
  - keeping snapshot: lxd-snapper20240617-220003
  - deleting snapshot: lxd-snapper20240616-220004 [ OK ]
  - deleting snapshot: lxd-snapper20240615-220010 [ OK ]
  - keeping snapshot: lxd-snapper20240531-220005
```

In July it looked better and backups were kept "correctly" (I guess the third weekly snapshot happens to fall in between the daily snapshot and therefore there's eight consecutive days?) until July 27:
```
  - keeping snapshot: lxd-snapper20240726-220006
  - keeping snapshot: lxd-snapper20240725-220005
  - keeping snapshot: lxd-snapper20240724-220013
  - keeping snapshot: lxd-snapper20240723-220006
  - keeping snapshot: lxd-snapper20240722-220005
  - keeping snapshot: lxd-snapper20240721-220004
  - keeping snapshot: lxd-snapper20240720-220010
  - keeping snapshot: lxd-snapper20240719-220005
  - deleting snapshot: lxd-snapper20240718-220004 [ OK ]
  - keeping snapshot: lxd-snapper20240713-220007
  - keeping snapshot: lxd-snapper20240706-220005
  - keeping snapshot: lxd-snapper20240630-220006
  - keeping snapshot: lxd-snapper20240531-220005
```

However, after that date stuff was deleted quickly. July 28 (still okay I guess, if 13, 20 and 27 are counted as weekly there's 3 weekly + 7 daily)
```
  - keeping snapshot: lxd-snapper20240727-220011
  - keeping snapshot: lxd-snapper20240726-220006
  - keeping snapshot: lxd-snapper20240725-220005
  - keeping snapshot: lxd-snapper20240724-220013
  - keeping snapshot: lxd-snapper20240723-220006
  - keeping snapshot: lxd-snapper20240722-220005
  - keeping snapshot: lxd-snapper20240721-220004
  - keeping snapshot: lxd-snapper20240720-220010
  - keeping snapshot: lxd-snapper20240719-220005
  - keeping snapshot: lxd-snapper20240713-220007
  - deleting snapshot: lxd-snapper20240706-220005 [ OK ]
  - keeping snapshot: lxd-snapper20240630-220006
  - keeping snapshot: lxd-snapper20240531-220005
```

July 29 (weekly snapshot 13 should not be deleted, this leaves only two weekly + 7 daily snapshots): 
```
  - keeping snapshot: lxd-snapper20240728-220004
  - keeping snapshot: lxd-snapper20240727-220004
  - keeping snapshot: lxd-snapper20240726-220004
  - keeping snapshot: lxd-snapper20240725-220003
  - keeping snapshot: lxd-snapper20240724-220010
  - keeping snapshot: lxd-snapper20240723-220003
  - keeping snapshot: lxd-snapper20240722-220004
  - keeping snapshot: lxd-snapper20240721-220003
  - keeping snapshot: lxd-snapper20240720-220007
  - deleting snapshot: lxd-snapper20240719-220003 [ OK ]
  - deleting snapshot: lxd-snapper20240713-220003 [ OK ]
  - keeping snapshot: lxd-snapper20240630-220003
  - keeping snapshot: lxd-snapper20240531-220002
```

July 30 (assuming 27 is the weekly snapshot, there are now 7 daily snapshots and only one weekly snapshot left):
```
  - keeping snapshot: lxd-snapper20240729-220005
  - keeping snapshot: lxd-snapper20240728-220005
  - keeping snapshot: lxd-snapper20240727-220011
  - keeping snapshot: lxd-snapper20240726-220006
  - keeping snapshot: lxd-snapper20240725-220005
  - keeping snapshot: lxd-snapper20240724-220013
  - keeping snapshot: lxd-snapper20240723-220006
  - keeping snapshot: lxd-snapper20240722-220005
  - deleting snapshot: lxd-snapper20240721-220004 [ OK ]
  - deleting snapshot: lxd-snapper20240720-220010 [ OK ]
  - keeping snapshot: lxd-snapper20240630-220006
  - keeping snapshot: lxd-snapper20240531-220005
```

July 31 keeps deleting (1 weekly + 6 daily):
```
  - keeping snapshot: lxd-snapper20240730-220004
  - keeping snapshot: lxd-snapper20240729-220004
  - keeping snapshot: lxd-snapper20240728-220004
  - keeping snapshot: lxd-snapper20240727-220004
  - keeping snapshot: lxd-snapper20240726-220004
  - keeping snapshot: lxd-snapper20240725-220003
  - keeping snapshot: lxd-snapper20240724-220010
  - deleting snapshot: lxd-snapper20240723-220003 [ OK ]
  - deleting snapshot: lxd-snapper20240722-220004 [ OK ]
  - keeping snapshot: lxd-snapper20240630-220003
  - keeping snapshot: lxd-snapper20240531-220002
```

And since then, every day a snapshot gets created and deleted (as if there is no more weekly snapshot but only daily ones. See yesterday's run (August 03):
```
  - keeping snapshot: lxd-snapper20240803-220003
  - keeping snapshot: lxd-snapper20240802-220003
  - keeping snapshot: lxd-snapper20240801-220004
  - keeping snapshot: lxd-snapper20240731-220003
  - keeping snapshot: lxd-snapper20240730-220004
  - keeping snapshot: lxd-snapper20240729-220004
  - keeping snapshot: lxd-snapper20240728-220004
  - deleting snapshot: lxd-snapper20240727-220004 [ OK ]
  - keeping snapshot: lxd-snapper20240630-220003
  - keeping snapshot: lxd-snapper20240531-220002
```

I suppose my issue is partly due to me not correctly understanding how snapshots are kept and when snapshots are considered "daily, weekly, monthly" (for example, I somehow expected that weekly snapshots are at least seven days old, but the behavior on July 27 seems to suggest that the current days backup can be considered weekly and cause the older ones to be deleted?).

However, I don't see three weekly backups right now on my instance, unless the monthly backups are somehow also counted as weekly (August 03). So maybe there is also some kind of bug in the pruning mechanism?

I just increased the `keep-weekly` amount to 5 to see how this changes things for this month. Unfortunately I'll have to wait until the end of the month until I can report back. Maybe someone has an idea in the meantime?

**Comment by Patryk27** [comment]
Hi,

Thanks for the detailed report! -- yes, it seems something's off in the logic here; I've got a hunch and I'm preparing a fix right now 🙂

**Comment by Patryk27** [comment]
Alright, I've just released v1.4.0 with the fixed algorithm.

**Comment by shaaati** [comment]
Thank you so much for your super quick response!
I deployed v1.4.0 and I'm looking forward to the end of this month, when everything should work :).
